### ISSUE TITLE
Consider running Vector as a separete user inside Docker

### ISSUE THREAD

**Issue by a-rodin** [timberio/vector#1599]
It was [proposed in the chat](https://gitter.im/timberio-vector/community?at=5e2acec8a420263e038ea69f) to use a dedicated non-root user in Vector's Docker containers, in order to improve security.

The [best practices for writing Dockerfiles](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user) recommend to create a separate user if the service inside the container can run without root privileges.

So I want to discuss here can all Vector's functions can be used if Vector is running inside of a container from a non-root user. In particular, we need to ensure that both `journald` and `docker` sources can be used to read logs on the host machine from within a container when the user inside the container is not root.

**Comment by binarylogic** [comment]
@ktff @LucioFranco I'm curious what you both think?

**Comment by LucioFranco** [comment]
Yes, we should definitely not be using the root user. As far as docker I think if the socket is available to the correct user it should work. I think this should be possible to set during build phase? I vaguely remember we will probably need to do some odd user stuff to set the correct permissions but I think it should be possible.

Looks like this thread shows a potential solution for docker https://forums.docker.com/t/mounting-using-var-run-docker-sock-in-a-container-not-running-as-root/34390/5

As for journald I am not totally sure I think @bruceg would know.

**Comment by ktff** [comment]
As @LucioFranco said, so long as socket is available, `docker_source` can work. 

Similar solution is to run Vector container with user in [`docker` group](https://docs.docker.com/install/linux/linux-postinstall/). [`docker run`](https://docs.docker.com/engine/reference/commandline/run/) accepts `--user` option

**Comment by binarylogic** [comment]
Sounds good. I'm going to work some on the Docker docs to include this as a best practice.

**Comment by LucioFranco** [comment]
The key thing is that we use the same environment variables as the `docker` cli tool uses to talk to the daemon, so if the user can get the `docker` cli to work then vector should work as well without any extra config.

**Comment by bruceg** [comment]
I completely agree with this, and was surprised to hear it wasn't set up that way to start with.

On my system, the systemd journal is configured (through ACLs) to be readable by any user in the `adm` or `wheel` groups. Permissions may be an issue depending on the user chosen and how the relevant GIDs are mapped in the container. Of course, similar issues apply to the `file` source, depending on the files it's reading from.

**Comment by binarylogic** [comment]
To update, I'll be addressing this in the guides work we're doing. Each component will link to a simple "getting started" guide for tha component and we  can address docker setup specifics in there.

**Comment by jszwedko** [comment]
This looks relevant for allowing users to configure the user from the outside: https://denibertovic.com/posts/handling-permissions-with-docker-volumes/

Seemingly this is necessary for sources that require access to volumes.

**Comment by jszwedko** [comment]
Hi @johanfagerstroem !

We started to investigate this, but have decided to just keep the image as root for now as running as non-root requires operators to be careful about running Vector when it needs access to host resources (commonly the case when deploying Vector as an agent or when using disk buffers) and so we anticipated a lot of confusion and frustration from Vector users. Instead, we plan to write some hardening docs for people that do want to run as non-root. As a note, you should be able to achieve this now by passing the user when running the Vector image (`--user` and `--group` when running via `docker`), you just have to make sure whatever user you run the image as has access to any needed resources.

The approach we were investigating for running as a `vector` user in the container required starting as root anyway and de-escalating as is common Docker images (see the ones in the library for example: https://github.com/docker-library) so a container scanning solution would still flag it anyway (as far as I know).

That all being said, we are open to more input here!

**Comment by nkinkade** [comment]
For what it's worth, I found that Vector needs read access to files that it might not have permission to read as a non-root user. Vector _seems_ happy running as root with all capabilities stripped except CAP_DAC_READ_SEARCH. This doesn't meet the condition of the original post, running the Vector container as non-root, but in this scenario root has only a single privileged capability (read files and list directories that it wouldn't otherwise have permission to do).

**Comment by jeidsath** [comment]
Hi @jszwedko 

>  Instead, we plan to write some hardening docs for people that do want to run as non-root

Is there a draft version of these documents anywhere?

**Comment by jszwedko** [comment]
> Hi @jszwedko
> 
> > Instead, we plan to write some hardening docs for people that do want to run as non-root
> 
> Is there a draft version of these documents anywhere?

There are some docs about hardening, https://vector.dev/docs/setup/going-to-prod/hardening/, but they don't cover running as non-root in Docker.

**Comment by uripe-wix** [comment]
Any news about this issue?

**Comment by pront** [comment]
Hello, I don't have the full context here but there are some linked issues documenting challenges and trade-offs. If someone from the community is willing to add a section [here](https://vector.dev/docs/setup/going-to-prod/hardening/) about running as a non-root user that would be a great start.

Workaround is documented here: https://github.com/vectordotdev/vector/issues/1599#issuecomment-1190318750

That being said, we want to fix this properly by running on a non-root user.
